File: sdk/noComment/HyperDbgDev/hyperdbg/hprdbgctrl/code/debugger/user-level/user-listening.cpp
#include "pch.h"
extern ACTIVE_DEBUGGING_PROCESS g_ActiveProcessDebuggingState;
extern DEBUGGER_SYNCRONIZATION_EVENTS_STATE
    g_UserSyncronizationObjectsHandleTable
        [DEBUGGER_MAXIMUM_SYNCRONIZATION_USER_DEBUGGER_OBJECTS];
VOID UdHandleUserDebuggerPausing(PDEBUGGEE_UD_PAUSED_PACKET PausePacket) {
  UdSetActiveDebuggingProcess(PausePacket->ProcessDebuggingToken,
                              PausePacket->ProcessId, PausePacket->ThreadId,
                              PausePacket->Is32Bit, TRUE);
  switch (PausePacket->PausingReason) {
  case DEBUGGEE_PAUSING_REASON_DEBUGGEE_ENTRY_POINT_REACHED:
    ShowMessages("reached to the entrypoint of the main module\n");
    break;
  case DEBUGGEE_PAUSING_REASON_DEBUGGEE_GENERAL_THREAD_INTERCEPTED:
    ShowMessages("\nthread: %x from process: %x intercepted\n",
                 PausePacket->ThreadId, PausePacket->ProcessId);
    break;
  default:
    break;
  }
  if (PausePacket->ReadInstructionLen != MAXIMUM_INSTR_SIZE) {
    if (HyperDbgLengthDisassemblerEngine(PausePacket->InstructionBytesOnRip,
                                         MAXIMUM_INSTR_SIZE,
                                         PausePacket->Is32Bit ? FALSE : TRUE) >
        PausePacket->ReadInstructionLen) {
      ShowMessages("oOh, no! there might be a misinterpretation in "
                   "disassembling the current instruction\n");
    }
  }
  if (!PausePacket->Is32Bit) {
    HyperDbgDisassembler64(PausePacket->InstructionBytesOnRip, PausePacket->Rip,
                           MAXIMUM_INSTR_SIZE, 1, TRUE,
                           (PRFLAGS)&PausePacket->Rflags);
  } else {
    HyperDbgDisassembler32(PausePacket->InstructionBytesOnRip, PausePacket->Rip,
                           MAXIMUM_INSTR_SIZE, 1, TRUE,
                           (PRFLAGS)&PausePacket->Rflags);
  }
  if (g_UserSyncronizationObjectsHandleTable
          [DEBUGGER_SYNCRONIZATION_OBJECT_USER_DEBUGGER_IS_DEBUGGER_RUNNING]
              .IsOnWaitingState == TRUE) {
    DbgReceivedUserResponse(
        DEBUGGER_SYNCRONIZATION_OBJECT_USER_DEBUGGER_IS_DEBUGGER_RUNNING);
  }
}

File: debugger/user_level/user_listening.go
package user_level

import (
	"fmt"

	"github.com/ddkwork/HyperDbg/debugger/misc"
	"github.com/ddkwork/golibrary/std/mylog"
	"github.com/ddkwork/sdk"
	"golang.org/x/sys/windows"
)

func HandleUserDebuggerPausing(
	pausePacket *sdk.DEBUGGEE_UD_PAUSED_PACKET,
	activeProcess *ActiveDebuggingProcess,
	syncObjects *[MaximumSyncObjects]SyncronizationEventState,
) {
	switch pausePacket.PausingReason {
	case sdk.DebuggeePausingReasonDebuggeeStartingModuleLoaded:
		mylog.Info("the target module is loaded and a breakpoint is set to the entrypoint")
		mylog.Info("press 'g' to reach to the entrypoint of the main module...")
	case sdk.DebuggeePausingReasonDebuggeeGeneralThreadIntercepted:
	default:
		break
	}

	if pausePacket.ReadInstructionLen != MaximumInstrSize {
		length, _ := misc.GetInstructionLength(pausePacket.InstructionBytesOnRip[:], func() misc.DisasmMode {
			if pausePacket.Is32Bit {
				return misc.Mode32
			}
			return misc.Mode64
		}())
		if length > uint32(pausePacket.ReadInstructionLen) {
			mylog.Warning("there might be a misinterpretation in disassembling the current instruction")
		}
	}

	if !activeProcess.IsPaused || pausePacket.ThreadId == activeProcess.ThreadId {
		activeProcess.IsActive = true
		activeProcess.ProcessId = pausePacket.ProcessId
		activeProcess.ThreadId = pausePacket.ThreadId
		activeProcess.Is32Bit = pausePacket.Is32Bit
		activeProcess.Rip = pausePacket.Rip
		activeProcess.ProcessDebuggingToken = pausePacket.ProcessDebuggingToken
		copy(activeProcess.InstructionBytesOnRip[:], pausePacket.InstructionBytesOnRip[:])
		activeProcess.IsPaused = true

		if !pausePacket.Is32Bit {
			fmt.Println()
			Disassemble64(pausePacket.InstructionBytesOnRip[:], pausePacket.Rip, MaximumInstrSize, 1, true, &pausePacket.Rflags)
		} else {
			fmt.Println()
			Disassemble32(pausePacket.InstructionBytesOnRip[:], pausePacket.Rip, MaximumInstrSize, 1, true, &pausePacket.Rflags)
		}
	}

	if syncObjects[SyncObjectUserDebuggerIsDebuggerRunning].IsOnWaitingState {
		windows.SetEvent(syncObjects[SyncObjectUserDebuggerIsDebuggerRunning].EventHandle)
	}
}

func Disassemble64(instructionBytes []byte, rip uint64, maxInstrSize, count uint32, showRflags bool, rflags *uint64) {
	for i := uint32(0); i < count && i*maxInstrSize < uint32(len(instructionBytes)); i++ {
		offset := i * maxInstrSize
		end := min(int(offset)+int(maxInstrSize), len(instructionBytes))
		inst, err := misc.DisassembleOne(instructionBytes[offset:end], rip+uint64(offset), misc.Mode64)
		if err != nil {
			continue
		}
		fmt.Printf("%s: %s\n", fmt.Sprintf("%x", inst.Address), misc.FormatDisassembly(inst))
	}
}

func Disassemble32(instructionBytes []byte, rip uint64, maxInstrSize, count uint32, showRflags bool, rflags *uint64) {
	for i := uint32(0); i < count && i*maxInstrSize < uint32(len(instructionBytes)); i++ {
		offset := i * maxInstrSize
		end := min(int(offset)+int(maxInstrSize), len(instructionBytes))
		inst, err := misc.DisassembleOne(instructionBytes[offset:end], rip+uint64(offset), misc.Mode32)
		if err != nil {
			continue
		}
		fmt.Printf("%s: %s\n", fmt.Sprintf("%x", inst.Address), misc.FormatDisassembly(inst))
	}
}
